python/src/main/python/drivers/graphicsfuzz_piglit_converter.py

```diff
@@ -0,0 +1,225 @@
+    shader_version_header = shader_lines[0]
+
+    shader_test_string += make_require_header(shader_version_header) + '\n'
+    shader_test_string += make_vertex_shader_header() + '\n'
+    shader_test_string += make_fragment_shader_header(shader_file_string) + '\n'
+    shader_test_string += make_test_header(shader_job_json_parsed)
+
+    return shader_test_string
+
+
+def make_require_header(shader_version_header: str) -> str:
+    """
+    Creates the piglit [require] header as well as the required GL and GLSL version strings.
+    Note: GLSL version strings are formatted as '#version ### (es)', where ### is a
+    specific GLSL version multiplied by 100, and (es) is an optional specifier that
+    determines whether to use GLES or not.
+    :param shader_version_header: the version string in the GLSL file.
+    :return: the shader_test require header string.
+    """
+    require_header = REQUIRE_HEADER + '\n'
+    # Piglit requires a version number with 1 digit of precision for the GL version, and
+    # 2 digits of precision for the GLSL version.
+    require_header += GLSLES_VERSION_STRING if ES_SPECIFIER in shader_version_header \
+        else GLSL_VERSION_STRING
+    try:
+        shader_version = shader_version_header.split(' ')[1]
+    except IndexError:
+        raise IOError('Malformed shader - invalid GLSL version string.')
+    if not shader_version.isdigit():
+        raise IOError('Malformed shader - invalid GLSL version string.')
+    require_header += format(float(shader_version) / SHADER_VERSION_FACTOR, '.2f') + '\n'
+    return require_header
+
+
+def make_vertex_shader_header() -> str:
+    """
+    Creates the piglit [vertex shader] header. Currently uses passthrough, but this function can
+    be modified later if we have a need for an explicit vertex shader.
+    :return: the shader_test vertex header string.
+    """
+    return VERTEX_HEADER + '\n'
+
+
+def make_fragment_shader_header(fragment_shader: str) -> str:
+    """
+    Creates the piglit [fragment shader] header.
+    :param fragment_shader: the fragment shader code.
+    :return: the shader_test fragment header string.
+    """
+    frag_header = FRAGMENT_HEADER + '\n'
+    frag_header += fragment_shader
+    return frag_header
+
+
+def make_test_header(shader_job_json_parsed: dict) -> str:
+    """
+    Creates the [test] header. Loads uniforms based on the uniforms found in the shader.
+    Note that piglit really doesn't like it if you try to load a uniform that isn't actually
+    used in the shader - because of this, we actually have to parse the shader itself and
+    find out which uniforms are being used instead of blindly pasting from the shader job JSON file.
+    :param shader_job_json_parsed: the parsed JSON properties.
+    :return: the shader_test test header string.
+    """
+    test_header = TEST_HEADER + '\n'
+    for key, value in shader_job_json_parsed.items():
+        test_header += UNIFORM_DEC + ' '
```

Comment:
This is somewhat subjective, but I prefer to use string templates for this kind of thing. This way, it is easy to see the overall structure of what is being output as opposed to many string concatenations.

```python
test_header += 'uniform {type} {uniform_name} {args}\n'.format(
    type=get_uniform_type_from_gl_func(value['func']),
    uniform_name=uniform_name,
    args=' '.join([str(arg) for arg in value['args']]),
)
```

I suggest renaming `key` to `uniform_name`. 

I also think it is fine to have the string template right here in the function because it will not be used anywhere else.

Reply:
I had no idea this was a thing - thanks.